### PR TITLE
Respawn

### PR DESCRIPTION
**Pull request by Thomas-Harit**
…th packets

### CODE REVIEW

File: cubic-server/Player.cpp [hunk 1]
```diff
@@ -1348,6 +1373,43 @@ void Player::teleport(const Vector3<double> &pos)
...
+        0, // Hashed seed
+        this->_gamemode, // Gamemode
+        this->_gamemode, // Previous gamemode
+        0, // Is debug
+        0, // Is flat
+        0, // Copy metadata
+        false // Has death location (need dimension name which we don't have yet)
+    });
+
+    for (auto &player : this->_dim->getPlayers()) {
```
Comment (huntears) [inline cubic-server/Player.cpp:1390]: Please don't get a reference to the players, as that would not increase the reference count to the shared_ptr and the server might just crash if a player gets deleted while you're working on it.

File: cubic-server/Player.cpp [hunk 2]
```diff
@@ -1348,6 +1373,43 @@ void Player::teleport(const Vector3<double> &pos)
...
+}
+
+void Player::kill()
+{
+    LivingEntity::kill();
+
+    this->sendCombatDeath({
+        this->_id, // Player id
+        0, // Killer Entity id (0 for now because we don't know him)
+        "You suck." // Message
```
Comment (huntears) [inline cubic-server/Player.cpp:1410]: Please use a better death message :)

File: cubic-server/Player.cpp [hunk 3]
```diff
@@ -1348,6 +1373,43 @@ void Player::teleport(const Vector3<double> &pos)
...
+    }
+}
+
+void Player::kill()
+{
+    LivingEntity::kill();
+
+    this->sendCombatDeath({
+        this->_id, // Player id
+        0, // Killer Entity id (0 for now because we don't know him)
```
Comment (huntears) [inline cubic-server/Player.cpp:1409]: Why do you not know him? I am pretty sure it is possible.

File: cubic-server/Player.hpp [hunk 4]
```diff
@@ -62,6 +62,7 @@ class Player : public LivingEntity {
     void setOperator(const bool isOp);
     void setKeepAliveId(long id);
     void updatePlayerInfo(const protocol::PlayerInfoUpdate &data);
+    void kill() override;
```
Comment (huntears) [inline cubic-server/Player.hpp:65]: Please document that method.

File: cubic-server/Player.hpp [hunk 5]
```diff
@@ -121,6 +122,7 @@ class Player : public LivingEntity {
     void sendSwingArm(bool mainHand, int32_t swingerId);
     void sendTeleportEntity(int32_t id, const Vector3<double> &pos, const Vector2<uint8_t> &rot);
     void sendRemoveEntities(const std::vector<int32_t> &entities);
+    void sendRespawn(const protocol::Respawn &data);
```
Comment (huntears) [inline cubic-server/Player.hpp:125]: Please document that method.

File: cubic-server/Player.hpp [hunk 6]
```diff
@@ -155,6 +157,7 @@ class Player : public LivingEntity {
     void sendDisplayObjective(const protocol::DisplayObjective &packet);
     void sendUpdateScore(const protocol::UpdateScore &packet);
     void sendUpdateTeams(const protocol::UpdateTeams &packet);
+    void sendCombatDeath(const protocol::CombatDeath &packet);
```
Comment (huntears) [inline cubic-server/Player.hpp:160]: Please document that method.

File: cubic-server/Player.hpp [hunk 7]
```diff
@@ -239,6 +243,9 @@ class Player : public LivingEntity {
     int _foodTickTimer;
     float _foodExhaustionLevel;
 
+    // Respawn
```
Comment (huntears) [inline cubic-server/Player.hpp:246]: That comment really is useless.

File: cubic-server/entities/LivingEntity.cpp [hunk 8]
```diff
@@ -38,6 +38,9 @@ void LivingEntity::damage(float damage)
     if (canceled)
         return;
     _health -= damage;
+    if (_health <= 0) {
+        this->kill();
```
Comment (huntears) [inline cubic-server/entities/LivingEntity.cpp:42]: You could pass the entity_id of the killer here :)

File: cubic-server/entities/LivingEntity.cpp [hunk 9]
```diff
@@ -67,6 +70,15 @@ void LivingEntity::knockback(const Vector3<double> &source, float force)
     }
 }
 
+/*
+ * @brief Kill the entity
+ */
```
Comment (huntears) [inline cubic-server/entities/LivingEntity.cpp:75]: Please don't document in cpp files, only do it in hpp :)

File: cubic-server/entities/LivingEntity.hpp [hunk 10]
```diff
@@ -28,6 +28,7 @@ class LivingEntity : public Entity {
     virtual void attack(const Vector3<double> &source);
     virtual void damage(float damage);
     virtual void knockback(const Vector3<double> &source = Vector3<double>(0, 0, 0), float force = KNOCKBACK_DEFAULT_FORCE);
+    virtual void kill();
```
Comment (huntears) [inline cubic-server/entities/LivingEntity.hpp:31]: Please document that method.

File: cubic-server/protocol/ClientPackets.hpp [hunk 11]
```diff
@@ -428,6 +430,13 @@ struct PlayerChatMessage {
 };
 std::unique_ptr<std::vector<uint8_t>> createPlayerChatMessage(const PlayerChatMessage &);
 
+struct CombatDeath {
+    int32_t playerId;
+    int32_t killerId;
+    chat::Message message;
+};
+std::unique_ptr<std::vector<uint8_t>> createCombatDeath(const CombatDeath &);
```
Comment (huntears) [inline cubic-server/protocol/ClientPackets.hpp:438]: Please document that function.

File: cubic-server/protocol/ClientPackets.hpp [hunk 12]
```diff
@@ -508,6 +517,21 @@ struct RemoveEntities {
...
+    player_attributes::Gamemode gamemode;
+    player_attributes::Gamemode previousGamemode;
+    bool isDebug;
+    bool isFlat;
+    bool copyMetadata;
+    bool hasDeathLocation;
+    std::string deathDimensionName;
+    Position deathLocation;
+};
+std::unique_ptr<std::vector<uint8_t>> createRespawn(const Respawn &in);
```
Comment (huntears) [inline cubic-server/protocol/ClientPackets.hpp:533]: Please document that function.

**Review by emneo-dev — CHANGES_REQUESTED** [review]
Quite a few things to change, I will be testing this PR tomorrow as I don't have the time to do it tonight.

**Comment by huntears** [comment]
Also please watch your commits names because `Fixed Merge` and `Fixed clang-format` are not proper commit messages

File: cubic-server/Player.cpp [hunk 13]
```diff
@@ -1348,6 +1373,43 @@ void Player::teleport(const Vector3<double> &pos)
     Entity::teleport(pos);
 }
 
+void Player::_respawn()
+{
+    this->sendRespawn({
+        "minecraft:overworld", // Dimension
+        "overworld", // World name
```
[the remark this answers is not in the document]
Comment (Trompettesib) [reply cubic-server/Player.cpp]: I'm not sure about that, cause you can respawn only in the overworld as we don't handle respawn anchors

File: cubic-server/Player.cpp [hunk 14]
```diff
@@ -1348,6 +1373,49 @@ void Player::teleport(const Vector3<double> &pos)
...
+}
+
+void Player::kill(const int32_t &killerId)
+{
+    LivingEntity::kill(killerId);
+
+    this->sendCombatDeath({
+        this->_id, // Player id
+        killerId, // Killer Entity id (0 for now because we don't know him)
+        "a better death message :)" // Message
```
Comment (huntears) [inline cubic-server/Player.cpp:1416]: :))))))

File: cubic-server/Player.hpp [hunk 15]
```diff
@@ -62,6 +62,7 @@ class Player : public LivingEntity {
     void setOperator(const bool isOp);
     void setKeepAliveId(long id);
     void updatePlayerInfo(const protocol::PlayerInfoUpdate &data);
+    void kill(const int32_t &killerId) override;
```
Comment (huntears) [inline cubic-server/Player.hpp:65]: Please document that method.

File: cubic-server/default/Overworld.hpp [hunk 16]
```diff
@@ -20,6 +20,9 @@ class Overworld : public Dimension {
     void stop() override;
     void generateChunk(int x, int z, world_storage::GenerationState goalState = world_storage::GenerationState::READY) override;
 
+    std::string getDimensionTypeName() const override { return "minecraft:overworld"; }
+    std::string getDimensionName() const override { return "overworld"; }
```
Comment (huntears) [inline cubic-server/default/Overworld.hpp:24]: Return const references here

**Review by emneo-dev — CHANGES_REQUESTED** [review]
Almost there :)

**Review by Trompettesib — COMMENTED** [review]
Looks good but segfault :/

**Review by emneo-dev — CHANGES_REQUESTED** [review]
Please don't invalidate other people's commit signatures if rebasing.
Only rebase when no conflicts are present, else do a merge then squash.
So please fix your commit history

File: cubic-server/Dimension.hpp [hunk 17]
```diff
@@ -132,6 +132,20 @@ class Dimension : public std::enable_shared_from_this<Dimension> {
      */
     [[nodiscard]] virtual world_storage::DimensionType getDimensionType() const { return _dimensionType; }
 
+    /**
+     * @brief Get the dimension type
+     *
+     * @return std::string dimension type
+     */
+    [[nodiscard]] virtual const std::string getDimensionTypeName() const = 0;
```
Comment (STMiki) [inline cubic-server/Dimension.hpp:140]: Use the `NODISCARD` define.
Will fix all of them later

**Review by STMiki — DISMISSED** [review]

**Review by STMiki — DISMISSED** [review]

**Review by STMiki — DISMISSED** [review]

**Review by emneo-dev — CHANGES_REQUESTED** [review]
image

Dead players are not sent to newly connected players, therefore they don't show up in the tab menu.

Just disconnect after killing someone then reconnect and you will not see their name appear in the tab menu.

**Review by Trompettesib — DISMISSED** [review]
Noice !

**Review by emneo-dev — CHANGES_REQUESTED** [review]
Please resolve the conflict with master before review.

**Pull request closed without merging**